File: src/doppler/examples/fir_demo.py
"""fir_demo.py — FIR filter demo for CF32 IQ signal streams.

Shows:
  1. Designing a low-pass filter with scipy and running it through doppler's
     FIR engine.
  2. State persistence across buffer boundaries.

Run:
  python examples/python/fir_demo.py
"""

import numpy as np
from scipy.signal import firwin

from doppler.filter import FIR

# ── 1. Design and apply a low-pass filter ────────────────────────────────────

N_TAPS = 15
CUTOFF = 0.1  # normalised cutoff (fraction of Nyquist)
N_BLOCK = 32

taps = firwin(N_TAPS, cutoff=CUTOFF, window="hann").astype(np.float32)
print("=== doppler FIR filter demo ===\n")
print(f"15-tap Hann-windowed sinc LP taps (cutoff = {CUTOFF} × Fs):")
print("  " + "  ".join(f"{t:7.4f}" for t in taps))
print()

fir = FIR(taps)

# Mixed input: 0.05 Fs tone (passband) + 0.4 Fs tone (stopband)
t_idx = np.arange(N_BLOCK)
lo_tone = np.exp(2j * np.pi * 0.05 * t_idx).astype(np.complex64)
hi_tone = (0.5 * np.exp(2j * np.pi * 0.4 * t_idx)).astype(np.complex64)
x = lo_tone + hi_tone

y = fir.execute(x)

print("--- CF32 input: 0.05 Fs + 0.4 Fs tone ---")
print("  first 8 samples (LF tone survives, HF attenuated):")
print(
    f"  {'idx':>3}  {'in.real':>8}  {'in.imag':>8}"
    f"  {'out.real':>9}  {'out.imag':>9}"
)
for i in range(8):
    print(
        f"  {i:>3}  {x[i].real:>8.3f}  {x[i].imag:>8.3f}"
        f"  {y[i].real:>9.3f}  {y[i].imag:>9.3f}"
    )
print()

# The engine must realise the designed response exactly: once the delay
# line fills, each complex tone emerges scaled by H(f) = Σ h[k]·e^{-j2πfk}.
k = np.arange(N_TAPS)
h_lo = complex(np.sum(taps * np.exp(-2j * np.pi * 0.05 * k)))
h_hi = complex(np.sum(taps * np.exp(-2j * np.pi * 0.4 * k)))
gain_lo_db = 20.0 * np.log10(abs(h_lo))
gain_hi_db = 20.0 * np.log10(abs(h_hi))
print(
    f"  designed gain: {gain_lo_db:+.1f} dB at 0.05 Fs (passband), "
    f"{gain_hi_db:+.1f} dB at 0.4 Fs (stopband)"
)
# LF tone survives (< 3 dB down), HF tone is crushed (> 55 dB down).
assert gain_lo_db > -3.0, f"passband gain {gain_lo_db:.1f} dB"
assert gain_hi_db < -55.0, f"stopband attenuation only {gain_hi_db:.1f} dB"
# Steady-state output matches the frequency-response prediction to CF32
# precision — this checks the actual C convolution, not just the design.
y_pred = h_lo * lo_tone + h_hi * hi_tone
steady = slice(N_TAPS - 1, None)
max_dev = float(np.max(np.abs(y[steady] - y_pred[steady])))
print(f"  steady-state max |y - H(f)·x| = {max_dev:.2e} — OK")
assert max_dev < 1e-4, f"output deviates from designed response: {max_dev}"
print()

# ── 2. State persistence across buffer boundaries ────────────────────────────

fir.reset()

t1 = np.arange(8)
t2 = np.arange(8, 16)


def tone(t):
    return np.exp(2j * np.pi * 0.05 * t).astype(np.complex64)


out1 = fir.execute(tone(t1)).copy()  # copy — execute returns a view
out2 = fir.execute(tone(t2))

print("--- stateful: 2×8 CF32 blocks (delay line persists) ---")
print(f"  block 1 last sample:  {out1[-1].real:.3f} + {out1[-1].imag:.3f}j")
print(f"  block 2 first sample: {out2[0].real:.3f} + {out2[0].imag:.3f}j")
print("  (continuity confirms delay line carries across calls)")
# The real test of state persistence: the two 8-sample blocks must equal
# a single 16-sample pass bit-for-bit — any dropped delay-line state
# would corrupt the first N_TAPS-1 samples of block 2.
fir_ref = FIR(taps)
ref = fir_ref.execute(tone(np.arange(16)))
assert np.array_equal(np.concatenate([out1, out2]), ref), (
    "split output differs from single-pass output — state lost"
)
print("  split 8+8 output == single 16-sample pass (bit-exact) — OK")
print()
print("Done.")

File: native/examples/fir_demo.c
/**
 * fir_demo.c — FIR filter demo for CF32 IQ signal streams.
 *
 * Shows:
 *   1. CF32 filtering — float IQ (GNU Radio compatible)
 *   2. State persistence across buffer boundaries
 *
 * The current API accepts only float complex (CF32) input.
 * Integer SDR input (CI8, CI16) must be converted to CF32 before
 * calling fir_execute().
 *
 * Build:
 *   make build
 *   ./build/native/examples/fir_demo
 */

#include <fir/fir_core.h>

#include <complex.h>
#include <math.h>
#include <stdio.h>
#include <stdlib.h>

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

/* Hann-windowed sinc low-pass filter, cutoff_norm in (0, 0.5]. */
static void
make_lowpass (float *taps, int N, double cutoff_norm)
{
  int half = N / 2;
  for (int k = 0; k < N; k++)
    {
      int    n    = k - half;
      double sinc = (n == 0) ? 1.0
                             : sin (M_PI * cutoff_norm * 2 * n)
                                   / (M_PI * cutoff_norm * 2 * n);
      double win  = 0.5 * (1.0 - cos (2.0 * M_PI * k / (N - 1)));
      taps[k]     = (float)(sinc * win);
    }
}

int
main (void)
{
  printf ("=== doppler FIR filter demo ===\n\n");

  /* 15-tap Hann-windowed sinc LP, cutoff = 0.1 * Fs */
  const int N = 15;
  float     taps[15];
  make_lowpass (taps, N, 0.1);

  printf ("15-tap Hann-windowed sinc LP taps:\n  ");
  for (int i = 0; i < N; i++)
    printf ("%7.4f", (double)taps[i]);
  printf ("\n\n");

  fir_state_t *fir = fir_create_real (taps, (size_t)N);
  if (!fir)
    {
      fprintf (stderr, "fir_create_real failed\n");
      return 1;
    }

  /* ------------------------------------------------------------------ *
   * 1. CF32 — complex tone at 0.05*Fs mixed with high-frequency noise  *
   * ------------------------------------------------------------------ */
  printf ("--- CF32 input (GNU Radio / native float) ---\n");
  {
    const int S = 32;
    float _Complex in[32], out[32];

    for (int i = 0; i < S; i++)
      {
        double phase_lo = 2.0 * M_PI * 0.05 * i; /* 0.05*Fs tone */
        double phase_hi = 2.0 * M_PI * 0.4 * i;  /* 0.4*Fs tone  */
        in[i] = CMPLXF ((float)(cos (phase_lo) + 0.5 * cos (phase_hi)),
                        (float)(sin (phase_lo) + 0.5 * sin (phase_hi)));
      }

    fir_execute (fir, in, S, out);

    printf ("  first 8 output samples (LF tone survives, HF attenuated):\n");
    for (int i = 0; i < 8; i++)
      printf ("  [%2d]  in=(%6.3f,%6.3f)  out=(%6.3f,%6.3f)\n", i,
              (double)crealf (in[i]), (double)cimagf (in[i]),
              (double)crealf (out[i]), (double)cimagf (out[i]));
    printf ("\n");
  }

  fir_reset (fir);

  /* ------------------------------------------------------------------ *
   * 2. State across calls — 2 blocks of 8 samples                      *
   * ------------------------------------------------------------------ */
  printf ("--- stateful: 2×8 CF32 blocks (delay line persists) ---\n");
  {
    float _Complex in1[8], in2[8], out1[8], out2[8];

    for (int i = 0; i < 8; i++)
      {
        double phase = 2.0 * M_PI * 0.05 * i;
        in1[i]       = CMPLXF ((float)cos (phase), (float)sin (phase));
      }
    for (int i = 0; i < 8; i++)
      {
        double phase = 2.0 * M_PI * 0.05 * (i + 8);
        in2[i]       = CMPLXF ((float)cos (phase), (float)sin (phase));
      }

    fir_execute (fir, in1, 8, out1);
    fir_execute (fir, in2, 8, out2);

    printf ("  block 1 last sample out:  (%6.3f, %6.3f)\n",
            (double)crealf (out1[7]), (double)cimagf (out1[7]));
    printf ("  block 2 first sample out: (%6.3f, %6.3f)\n",
            (double)crealf (out2[0]), (double)cimagf (out2[0]));
    printf ("  (continuity confirms delay line carries across calls)\n\n");
  }

  fir_destroy (fir);
  printf ("Done.\n");
  return 0;
}
